From 637ea0c5931e4c904828be90941eacb431d9b2e1 Mon Sep 17 00:00:00 2001
From: Wei-Chiu Chuang <weichiu@cloudera.com>
Date: Mon, 2 May 2016 06:43:50 -0700
Subject: [PATCH 1476/2748] HDFS-10319. Balancer should not try to pair
 storages with different types.

(cherry picked from commit e7f1b8fcb35f8c9530a00ca0475bdd308003e5c8)

Change-Id: I2c22ee256f751bcfecd6d58712a8f81f2ffea86e
---
 .../hadoop/hdfs/server/balancer/Balancer.java      |   10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/balancer/Balancer.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/balancer/Balancer.java
index d86b56c..ac51e66 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/balancer/Balancer.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/balancer/Balancer.java
@@ -467,8 +467,7 @@ C chooseCandidate(G g, Iterator<C> candidates, Matcher matcher) {
         final C c = candidates.next();
         if (!c.hasSpaceForScheduling()) {
           candidates.remove();
-        } else if (matcher.match(dispatcher.getCluster(),
-            g.getDatanodeInfo(), c.getDatanodeInfo())) {
+        } else if (matchStorageGroups(c, g, matcher)) {
           return c;
         }
       }
@@ -476,6 +475,13 @@ C chooseCandidate(G g, Iterator<C> candidates, Matcher matcher) {
     return null;
   }
 
+  private boolean matchStorageGroups(StorageGroup left, StorageGroup right,
+      Matcher matcher) {
+    return left.getStorageType() == right.getStorageType()
+        && matcher.match(dispatcher.getCluster(),
+            left.getDatanodeInfo(), right.getDatanodeInfo());
+  }
+
   /* reset all fields in a balancer preparing for the next iteration */
   void resetData(Configuration conf) {
     this.overUtilized.clear();
-- 
1.7.9.5

